### PR TITLE
Added gravity as attribute to Entities

### PR DESCRIPTION
**Pull request by mcenderdragon**
## Why Attributes and not Events:
As sending an event each tick is critical enough sending each tick at least 1 packet for each entity seems a bit overkill. For 100 entries this would mean 2000 events per second and 120000 events per minute. If the performance is still good enough, the GC will quickly get overloaded considering what else is also going on when mods are installed.

Attributes are already a good vanilla concept of easily modifiable entity-bound values, so why reinventing the wheel?.

## What do we gain from this?
While it is already possible to make something like a reduced gravity effect, by applying motion each tick to the entity it does not look exactly that good and is not performance effective for larger groups.
We can now do things like biome specific or dimension specific gravity, asteroid belts would be easily possible. It is also possible to make slow fall boots, or changing "feather falling" to actually fall more slowly.

## TODO:
- [x] get the test mod running (it loads in my def env, because I put i in the main sourceSet)
- [x] Attribute needs only to change on the server and it will update on the client correctly!
- [x] Jump height updates correctly
- [x] check if potions can apply modifiers
- [x] check if armor can apply modifiers

### CODE REVIEW

**Comment by Alpvax** [comment]
Possibly should be called fall speed, as the gravity itself would be due to the biome/dimension.
Alternatively, we could make it affect jump height as well (I can't tell whether it already does as I'm on my phone, so if so ignore me).

**Comment by mcenderdragon** [comment]
How does the jump boost potion work? In theorie a jump should be higher when the down-force is lower, but I will test this.

**Comment by mcenderdragon** [comment]
I can confirm jump height changes accordingly, so no extra afford from modders is needed here.

**Comment by mcenderdragon** [comment]
Currently even when you  fall slowly the falll distance is keept, so you will get the "old/normal" fall damage. I dont want to change that field as I am sure other mods need the correct value. So what are other opinions on this? 
We have the LivingFallEvent so this should be no issue.

**Comment by mcenderdragon** [comment]
I added a gravity stick to the test mod, which increseaes the gravty when held. The modifiers of the stick & the biome modifiers correctly stack together. I did not test potions but as items work they will also work. 
@mezz & @tterrag1098, this PR is ready for a review.

**Comment by Alpvax** [comment]
I thought fall damage was modified to be velocity based instead of just distance based. Was that a mod-added change?

**Comment by mcenderdragon** [comment]
it could be both, but what I see codewise is all relative to fallDistance (which is set constantly to 1 by vanillas slow fall potion)

**Comment by mcenderdragon** [comment]
Fixed merge confilcts, @tterrag1098 & @mezz this is ready for a review.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 1]
```diff
@@ -1,24 +1,29 @@
...
++   private static final UUID SLOW_FALLING_ID = UUID.fromString("A5B6CF2A-2F7C-31EF-9022-7C3E7D5E6ABA");
     private static final AttributeModifier field_110157_c = (new AttributeModifier(field_110156_b, "Sprinting speed boost", (double)0.3F, 2)).func_111168_a(false);
++   private static final AttributeModifier SLOW_FALLING = new AttributeModifier(SLOW_FALLING_ID, "Slow falling aceleration reduction", -0.07, 0).func_111168_a(false); // Add -0.07 to 0.08 so we get the vanilla default of 0.01
 +   public static final net.minecraft.entity.ai.attributes.IAttribute SWIM_SPEED = new net.minecraft.entity.ai.attributes.RangedAttribute(null, "forge.swimSpeed", 1.0D, 0.0D, 1024.0D).func_111112_a(true);
 +   public static final net.minecraft.entity.ai.attributes.IAttribute NAMETAG_DISTANCE = new net.minecraft.entity.ai.attributes.RangedAttribute(null, "forge.nameTagDistance", 64.0D, 0.0D, Float.MAX_VALUE).func_111112_a(true);
++   public static final net.minecraft.entity.ai.attributes.IAttribute ENTITY_GRAVITY = new net.minecraft.entity.ai.attributes.RangedAttribute(null, "forge.entity_gravity", 0.08D, -8.0D, 8-0D).func_111112_a(true);
```
Comment (mezz) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: `8-0D` ?

File: src/test/java/net/minecraftforge/debug/entity/GravityAttributeTest.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        public ItemGravityStick(Properties properties) 
+        {
+            super(properties);
+        }
+        
+        @Override
+        public Multimap<String, AttributeModifier> getAttributeModifiers(EntityEquipmentSlot slot) 
+        {
+            @SuppressWarnings("deprecation")
+			Multimap<String, AttributeModifier> multimap = super.getAttributeModifiers(slot);
```
Comment (mezz) [inline src/test/java/net/minecraftforge/debug/entity/GravityAttributeTest.java]: please fix indentation

File: src/test/java/net/minecraftforge/debug/entity/GravityAttributeTest.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            FMLJavaModLoadingContext.get().getModEventBus().register(this);
+        }
+    }
+
+    @SubscribeEvent
+    public void worldTick(TickEvent.WorldTickEvent event)
+    {
+        if (ENABLE && !event.world.isRemote )
+        {
+            if(ticks++ > 60)
```
Comment (mezz) [inline src/test/java/net/minecraftforge/debug/entity/GravityAttributeTest.java]: Please fix all formatting in this file. for example:
```suggestion
            if (ticks++ > 60)
```

File: src/test/java/net/minecraftforge/debug/entity/GravityAttributeTest.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            logger = LogManager.getLogger();
+            MinecraftForge.EVENT_BUS.register(this);
+            FMLJavaModLoadingContext.get().getModEventBus().register(this);
+        }
+    }
+
+    @SubscribeEvent
+    public void worldTick(TickEvent.WorldTickEvent event)
+    {
+        if (ENABLE && !event.world.isRemote )
```
Comment (mezz) [inline src/test/java/net/minecraftforge/debug/entity/GravityAttributeTest.java]: Please fix all formatting in this file. for example:
```suggestion
        if (ENABLE && !event.world.isRemote)
```

**Review by mezz — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch (hunk 1 above)
Comment (mcenderdragon) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: fixed typo

File: src/test/java/net/minecraftforge/debug/entity/GravityAttributeTest.java (hunk 4 above)
Comment (mcenderdragon) [reply src/test/java/net/minecraftforge/debug/entity/GravityAttributeTest.java]: I hope I did get all of these mistakes.

**Review by mezz — APPROVED** [review]

**Comment by mcenderdragon** [comment]
The test mod is confirmed as working. 
(I had to run via eclipse forge_test_client)
@mezz I hope I did not oversee formating issues again.

**Comment by mezz** [comment]
Please update to resolve conflicts

**Comment by mcenderdragon** [comment]
Merge conflicts have been resolved.

File: patches/minecraft/net/minecraft/entity/LivingEntity.java.patch [hunk 5]
```diff
@@ -1,24 +1,29 @@
 --- a/net/minecraft/entity/LivingEntity.java
 +++ b/net/minecraft/entity/LivingEntity.java
-@@ -105,6 +105,8 @@
+@@ -104,7 +104,12 @@
+ 
  public abstract class LivingEntity extends Entity {
     private static final UUID field_110156_b = UUID.fromString("662A6B8D-DA3E-4C1C-8813-96EA6097278D");
++   private static final UUID SLOW_FALLING_ID = UUID.fromString("A5B6CF2A-2F7C-31EF-9022-7C3E7D5E6ABA");
     private static final AttributeModifier field_110157_c = (new AttributeModifier(field_110156_b, "Sprinting speed boost", (double)0.3F, AttributeModifier.Operation.MULTIPLY_TOTAL)).func_111168_a(false);
++   private static final AttributeModifier SLOW_FALLING = new AttributeModifier(SLOW_FALLING_ID, "Slow falling aceleration reduction", -0.07, AttributeModifier.Operation.ADDITION).func_111168_a(false); // Add -0.07 to 0.08 so we get the vanilla default of 0.01
```
Comment (HellFirePvP) [inline patches/minecraft/net/minecraft/entity/LivingEntity.java.patch]:
```suggestion
+   private static final AttributeModifier SLOW_FALLING = new AttributeModifier(SLOW_FALLING_ID, "Slow falling acceleration reduction", -0.07, AttributeModifier.Operation.ADDITION).func_111168_a(false); // Add -0.07 to 0.08 so we get the vanilla default of 0.01
```
Fixes typo

**Comment by mcenderdragon** [comment]
@gigaherz all merge conflicts are solved, so according to lex (discord) this PR is ready.

**Pull request merged**